<a href="https://colab.research.google.com/github/LimGuenTaek/CLRS/blob/master/01_%ED%8C%8C%EC%9D%B4%EC%8D%AC%20%EA%B8%B0%EB%B0%98%EC%9D%98%20%EB%A8%B8%EC%8B%A0%EB%9F%AC%EB%8B%9D%EA%B3%BC%20%EC%83%9D%ED%83%9C%EA%B3%84%20%EC%9D%B4%ED%95%B4%20/1_4_%ED%8C%90%EB%8B%A4%EC%8A%A4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 데이터 핸들링 - 판다스

* 판다스는 파이썬에서 데이터 처리를 위해 존재하는 가장 인기 있는 라이브러리입니다.

* 일반적으로 대부분의 데이터 세트는 2차원 데이터 입니다. 즉 , Row x Column으로 구성되어 있는데

* 2차원 데이터가 인기 있는 이유는 바로 인간이 가장 이해하기 쉬운 데이터 구조이면서도 효과적으로 데이터를 담을 수 있는 구조이기 때문입니다.

* 판다스가 파이썬 세계의 대표적인 데이터 핸들링 프레임 워크이지만 , 광범위한 영역을 커버하고 있기에 이를 마스터하기에는 의외로 많은 노력과 시간 투자가 필요합니다.

* 판다스의 핵심 객체는 DataFrame 입니다. DataFrame 은 여러 개의 행과 열로 이뤄진 2차원 데이터를 담는 데이터 구조체입니다.

* DataFrame을 이해하기 전에 다른 중요 객체인 Index와 Series를 이해하는 것도 중요합니다.

* Index는 개별 데이터를 고유하게 식별하는 Key값이고 , Series는 Column이 하나뿐인 데이터 구조체입니다.

* 판다스는 csv , tab과 같은 다양한 유형의 분리 문자로 칼럼을 분리한 파일을 손쉽게 DataFrame으로 로딩할 수 있게 해줍니다.

## 1. 판다스 시작 - 파일을 DataFrame으로 로딩 , 기본 API 

In [1]:
import pandas as pd

* 판다스는 다양한 포맷으로 된 파일을 DataFrame으로 로딩할 수 있는 편리한 API를 제공합니다.

* 대표적으로 read_csv() , read_table() , read_fwf()가 있습니다.

* read_csv()는 이름에서도 알 수 있듯이 csv(칼럼을 ','으로 구분한 파일 포맷) 파일 포맷 변환을 위한 API입니다.

* read_table() , read_csv()의 가장 큰 차이는 필드 구분 문자(Delimeter)r가 ',' 냐 '\t' 이냐의 차이입니다.

**read_csv(filepath or buffer , sep=',')**

* 함수에서 가장 중요한 인자는 filepath입니다. 나머지 인자는 지정하지 않으면 default 값으로 할당됨

* DataFrame 객체를 잠시 살펴보면 데이터 파일의 첫 번째 줄에 있던 칼럼 문자열이 DataFrame의 칼럼으로 할당 됐습니다. 

* read_csv()는 별다른 파라미터 지정이 없으면 파일의 맨 처음 로우를 칼럼명으로 인지하고 칼럼으로 변환합니다.

* 칼럼 명이 표시되어 있지 않은 숫자들이 판다스의 Index객체 값입니다.

In [2]:
train_data=pd.read_csv('/content/drive/MyDrive/Python_Machine_Learning/titanic/train.csv')
train_data.head(3) # 앞에서부터 3개의 row만 출력 , default는 5개

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S

[3 rows x 12 columns]

In [3]:
print("csv 파일의 type : {}".format(type(train_data)))
print("csv 파일의 shape : {}".format(train_data.shape))

csv 파일의 type : <class 'pandas.core.frame.DataFrame'>
csv 파일의 shape : (891, 12)


* info() 메소드를 사용하면 Column의 타입 , Null 데이터 개수 , 데이터 분포도 , 메타 데이터등의 정보를 조회할 수 있습니다.

* describe() 메소드를 사용하면 칼럼별 숫자형 데이터값의 n-percentile 분포도 , 평균값 , 최댓값 , 최솟값을 나타냅니다. 오직 숫자형(int , float 등) 칼럼의 분포도만 ㅈ사하며 자동으로 object 타입의 칼럼은 출력에서 제외시킵니다.

* 데이터의 분포도를 아는 것은 머신러닝 알고리즘의 성능을 향상시키는 중요한 요소입니다.

* 가령 회귀에서 결정 값이 정규 분포를 이루지 않고 특정 값으로 왜곡돼 있는 경우 , 또는 데이터값에 이상치가 많을 경우 예측 성능이 저하됩니다.

* describe() 메소드만으로 정확한 분포도를 알기에는 무리가 있겠지만 , 개략적인 수준의 분포도를 확인 할 수 있어 유용합니다.

In [4]:
train_data.info()
print('\n\n')
train_data.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB





PassengerId    Survived      Pclass  ...       SibSp       Parch        Fare
count   891.000000  891.000000  891.000000  ...  891.000000  891.000000  891.000000
mean    446.000000    0.383838    2.308642  ...    0.523008    0.381594   32.204208
std     257.353842    0.486592    0.836071  ...    1.102743    0.806057   49.693429
min       1.000000    0.000000    1.000000  ...    0.000000    0.000000    0.000000
25%     223.500000    0.000000    2.000000  ...    0.000000    0.000000    7.910400
50%     446.000000    0.000000    3.000000  ...    0.000000    0.000000   14.454200
75%     668.500000    1.000000    3.000000  ...    1.000000    0.000000   31.000000
max     891.000000    1.000000    3.000000  ...    8.000000    6.000000  512.329200

[8 rows x 7 columns]

* count는 Not Null인 데이터 건수 , mean은 전체 데이터의 평균값 , std는 표준편차 , min은 최솟값, max는 최댓값입니다.

* 또한 describe()는 해당 숫자 칼럼이 숫자형 카테고리 칼럼인지를 판단할 수 있게 도와줍니다.

* PassengerID 칼럼은 승객 ID를 식별하는 칼럼이므로 1~891까지 숫자가 할당되어서 분석을 위한 의미 있는 속성이 아닙니다.

* Survived의 경우 min 0 , 25%~75%도 0 , max가 1이므로 0,1로 이루어진 숫자형 카테고리 칼럼입니다.

* Pclass의 경우도 숫자 1,2,3으로 이루어진 숫자형 카테고리 칼럼인데 Pclass 칼럼이 어떠한 분포로 구성되어 있는지 살펴보겠습니다.

* Dataframe의 [  ] 연산자 내부에 칼럼명을 입력하면 Series 형태로 특정 칼럼 데이터 세트가 반환됩니다.

* 이렇게 반환된 Series 객체에 value_counts() 메소드를 호출하면 해당 칼럼값의 유형과 건수를 확인할 수 있습니다.

* value_counts()는 지정된 칼럼의 데이터값 건수를 반환합니다. 이는 데이터의 분포도를 확인하는 데 매우 유용한 함수이며 , 앞으로 자주 사용되므로 진해지면 좋을 것입니다.

* value_counts()는 DataFrame에서는 사용할 수 없습니다.

In [5]:
value=train_data['Pclass']
print(type(value),value.shape)
print(value.value_counts()) # Data 별로 분포도를 나타내줌

<class 'pandas.core.series.Series'> (891,)
3    491
1    216
2    184
Name: Pclass, dtype: int64


## 2. DataFrame과 리스트 , 딕셔너리 , 넘파이 ndarray 상호 변환

* 앞에서 csv파일을 read_csv()를 이용해 DataFrame으로 생성했습니다.

* 기본적으로 DataFrame은 파이썬의 리스트 , 딕셔너리 그리고 넘파이 ndarray 등 다양한 데이터로부터 생성될 수 있습니다.

* 또한 DataFrame은 반대로 리스트 , 딕셔너리 , 넘파이 ndarray 등으로 변환될 수 있습니다.

* 특히 사이킷런의 많은 API는 DataFrame을 인자로 입력 받을 수 있지만 , 넘파이 ndarray를 입력 인자로 사용하는 경우가 대부분입니다.

* 따라서 DataFrame과 넘파이 ndarray 상호 간의 변환은 매우 빈번하게 발생합니다.

**넘파이 ndarray , 파이썬 리스트 , 딕셔너리로부터 DataFrame을 생성해보겠습니다.**

* 일반적으로 DataFrame으로 변환 시 칼럼명을 지정해 줍니다.(저장해주지 않으면 자동으로 칼럼명을 할당합니다.)

* 판단스 DataFrame 객체의 생성 인자 data는 리스트나 딕셔너리 또는 넘파이 ndarray를 입력 받고 , 생성 인자 columns는 칼럼명 리스트를 입력 받아서 쉽게 DataFrame을 생성할 수 있습니다.

* DataFrame은 기본적으로 행과 열을 가지는 2차원 데이터 입니다. 따라서 2차원 이하의 데이터들만 DataFrame으로 변환 될 수 있습니다.


In [6]:
import numpy as np
col_name=['col1','col2','col3']
list2=[[1,2,3],
       [11,22,33]]

array2=np.array2=[[1,2,3],
           [11,22,33]]

df_list2=pd.DataFrame(list2,columns=col_name)
df_array2=pd.DataFrame(array2,columns=col_name)

print(df_list2)
print()
print(df_array2)


   col1  col2  col3
0     1     2     3
1    11    22    33

   col1  col2  col3
0     1     2     3
1    11    22    33


* 이번에는 딕셔너리를 DataFrame으로 변환해 보겠습니다.

* 딕셔너리의 key는 칼럼명으로 value는 칼럼데이터로 변환됩니다.

In [7]:
dict={'col1':[1,11],'col2':[2,22],'col3':[3,33]}
df_dict=pd.DataFrame(dict)
print(df_dict)

   col1  col2  col3
0     1     2     3
1    11    22    33


**DataFrame을 넘파이 ndarray,리스트,딕셔너리로 변환하기**

* 많은 머신러닝 패키지가 기본 데이터 형으로 넘파이 ndarray를 사용합니다.

* 따라서 데이터 핸들링은 DataFrame을 이용하더라도 머신러닝 패키지의 입력 인자 등에 적용하기 위해 다시 넘파이 ndarray로 변환하는 경우가 빈번하게 발생합니다.

* DataFrame을 넘파이 ndarray로 변환하는 것은 DataFrame 객체의 values를 이용해 쉽게 할 수 있습니다.

* 이 책 전반에 걸쳐서 values를 이용한 ndarray로의 변환은 매우 많이 사용되므로 반드시 기억해둘 것

In [8]:
# DataFrame을 넘파이array로 변환
array3=df_dict.values
print(type(array3),array3.shape)
print(array3,end='\n\n')

#DataFrame을 리스트로 변환
list3=df_dict.values.tolist()
print(list3,end='\n\n')

#DataFrame을 딕셔너리로 변환
dict3=df_dict.to_dict('list')
print(dict3,end='\n\n')

<class 'numpy.ndarray'> (2, 3)
[[ 1  2  3]
 [11 22 33]]

[[1, 2, 3], [11, 22, 33]]

{'col1': [1, 11], 'col2': [2, 22], 'col3': [3, 33]}



## 3. DataFrame의 칼럼 데이터 세트 생성과 수정

* DataFrame의 칼럼 데이터 세트 생성과 수정 역시 [  ] 연산자를 이용해 쉽게 할 수 있습니다.

* 먼저 Titanic DataFrame에 새로운 칼럼 Age_0 을 추가하고 일괄적으로 0 값을 할당하겠습니다.

* train_data['Age_0']=0과 같이 series에 상수값을 할당하면 Series의 모든 데이터 세트에 일괄적으로 적용됩니다.

* 이는 넘파이의 ndarray에 상수값을 할당할 때 모든 ndarray 값에 일괄 적용됨과 동일합니다.


In [9]:
train_data['Age_0']=0
train_data.head(3)

PassengerId  Survived  Pclass  ... Cabin Embarked  Age_0
0            1         0       3  ...   NaN        S      0
1            2         1       1  ...   C85        C      0
2            3         1       3  ...   NaN        S      0

[3 rows x 13 columns]

In [10]:
train_data['Age_by_10']=train_data['Age']*10
train_data.head(5)

PassengerId  Survived  Pclass  ... Embarked Age_0  Age_by_10
0            1         0       3  ...        S     0      220.0
1            2         1       1  ...        C     0      380.0
2            3         1       3  ...        S     0      260.0
3            4         1       1  ...        S     0      350.0
4            5         0       3  ...        S     0      350.0

[5 rows x 14 columns]

## 4. DataFrame 데이터 삭제

* DataFrame에서 데이터의 삭제는 drop() 메소드를 이용합니다.

* drop()메소드는 사용하는데 혼동을 느낄만한 부분이 있어 주의가 필요합니다.

    > DataFrame.drop(**labels=None , axis=0**,index=None,columns=None,level=None,**inplace=False**,errors='raise')

* 이중 가장 중요한 파라미터는 labels , axis , inplace 입니다.

* 먼저 axis 값에 따라서 특정 칼럼 또는 특정 행을 drop합니다.(axis 0은 로우 방향 , axis 1은 칼럼 방향 입니다.)

* 따라서 axis=1을 입력하면 칼럼 축 방향으로 drop을 수행하므로 칼럼을 드롭하겠다는 의미입니다.

* labels에 원하는 칼럼 명을 입력하고 axis=1을 입력하면 지정된 칼럼을 드롭합니다.

* axis=0으로 입력하면 row를 삭제한다는 의미이고 주로 이상치 데이터를 삭제해줄 때 사용합니다.

In [11]:
# Age_0 column 삭제
train_drop=train_data.drop('Age_0',axis=1)
train_drop.head(3)

PassengerId  Survived  Pclass  ... Cabin Embarked  Age_by_10
0            1         0       3  ...   NaN        S      220.0
1            2         1       1  ...   C85        C      380.0
2            3         1       3  ...   NaN        S      260.0

[3 rows x 13 columns]

* 이번에는 또 하나의 주요 파라미터인 inplace에 대해 알아보겠습니다.

* inplace가 False로 되어있으면 자기 자신의 DataFrame을 삭제 하지 않으며 , 삭제된 결과 DataFrame을 반환합니다.

* inplace가 True로 되어있으면 자신의 DataFrame의 데이터를 삭제합니다.

In [12]:
train_drop=train_data.drop(['Age_0','Age_by_10'],axis=1,inplace=True)
train_data.head(3)

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S

[3 rows x 12 columns]

## 5. Index 객체

* 판다스의 Index 객체는 DataFrame , Series의 레코드를 고유하게 식별하는 객체입니다.

* DataFrame , Series 에서 Index 객체만 추출하려면 DataFrame.index , Series.index 속성을 통해 가능합니다.

In [13]:
train_data=pd.read_csv('/content/drive/MyDrive/Python_Machine_Learning/titanic/train.csv')
indexes=train_data.index
print(indexes)
#print(indexes.values)
print(type(indexes.values))
print(indexes.values.shape)

RangeIndex(start=0, stop=891, step=1)
<class 'numpy.ndarray'>
(891,)


* 이때 , 한번 만들어진 DataFrame 및 series의 Index 객체는 함부로 변경할 수 없습니다.

* 즉 다음과 같이 첫 번째 Index 객체의 값을 5로 변경하는 작업은 수행할 수 없습니다.

In [14]:
##indexes[0]=5 에러가 날 것임

* DataFrame 및 Series에 reset_index() 메소드를 수행하면 새롭게 인덱스를 연속 숫자 형으로 할당하며 기존 인덱스는 'index'라는 새로운 칼럼 명으로 추가합니다.

* reset_index()는 인덱스가 연속된 int 숫자형 데이터가 아닐 경우에 다시 이를 연속 int 숫자형 데이터로 만들 때 주로 사용합니다.

* Series에 reset_index()를 적용해주면 DataFrame이 반환됩니다.(Column이 두개가 되기 때문)

* reset_index()의 parameter 중 drop=True로 설정하면 기존 인덱스는 새로운 칼럼으로 추가되지 않고 삭제(drop)됩니다.

In [16]:
train_reset_index=train_data.reset_index(inplace=False)
train_reset_index.head(3)

index  PassengerId  Survived  ...     Fare Cabin Embarked
0      0            1         0  ...   7.2500   NaN        S
1      1            2         1  ...  71.2833   C85        C
2      2            3         1  ...   7.9250   NaN        S

[3 rows x 13 columns]

## 6. 데이터 셀렉션 및 필터링

* 판다스의 데이터 셀렉션과 필터링은 넘파이와 상당히 유사한 부분도 있고 다른 부분도 있기에 사용할때마다 혼동하기 쉽습니다.

* 판다스의 DataFrame과 Series끼리도 이러한 데이터 셀렉션 기능이 달라지는 부분이 있어 주의가 필요합니다.

* 넘파이의 경우 [  ] 연산자내 단일 값 추출 , 슬라이싱 , 팬시 인덱싱 , 불린 인덱싱을 통해 데이터를 추출했습니다.

* 판다스의 경우 ix[ ] , iloc[ ] , loc[ ] 연산자를 통해 동일한 작업을 수행합니다.

* ix , iloc , loc을 접하기 전에 먼저 판다스의 [ ] 연산자가 넘파이의 [ ]와 어떠한 차이가 있는지 알아보겠습니다.

**DataFrame의 [ ] 연산자**

* 우선 , DataFrame 바로 뒤에 있는 [  ] 안에 들어갈 수 있는 것은 칼럼 명 문자 또는 인덱스로 변환 가능한 표현식입니다.

* 현재 수준에서는 DataFrame 뒤에 있는 [  ]는 칼럼만 지정할 수 있는 '칼럼 지정 연산자'로 이해하는 게 혼돈을 막는 가장 좋은 방법입니다.

In [22]:
print(train_data['Pclass'].head())
print()
print(train_data[['Survived','Pclass']].head())

0    3
1    1
2    3
3    1
4    3
Name: Pclass, dtype: int64

   Survived  Pclass
0         0       3
1         1       1
2         1       3
3         1       1
4         0       3


* DataFrame의 [  ]내에 숫자 값을 입력할 경우 오류가 발생한다고 했는데 , 판다스의 인덱스 형태로 변환 가능한 표현식은 [  ]내에 입력할 수 있습니다.

In [27]:
train_data[0:2]

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C

[2 rows x 12 columns]

In [29]:
train_data[train_data['Pclass']==3].head(5)

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
2            3         1       3  ...   7.9250   NaN         S
4            5         0       3  ...   8.0500   NaN         S
5            6         0       3  ...   8.4583   NaN         Q
7            8         0       3  ...  21.0750   NaN         S

[5 rows x 12 columns]

* DataFrame 뒤의 [  ]연산자의 입력 인자가 여러 가지 형태이므로 사용할 때 꽤 혼동을 줍니다. 혼돈을 방지하귀 위한 좋은 가이드는 다음과 같습니다.

  * DataFrame 바로 뒤의 [    ]연산자는 넘파이의 [    ]나 Series[ ]와 다릅니다.
  * DataFrame 바로 뒤의 [ ]내 입력 값은 칼럼명(또는 칼럼의 리스트)을 지정해 칼럼 지정 연산에 사용하거나 불린 인덱스 용도로만 사용해야 합니다.
  * DataFrame[0:2]와 같은 슬라이싱 연산으로 데이터를 추출하는 방법은 사용하지 않는 게 좋습니다.

**DataFrame ix[ ] 연산자**

* ix[ ]는 ix[0 , 'Pclass']와 같이 행 위치 지정으로 인덱스 값 0 , 칼럼 명 'Pclass' 를 입력해 원하는 위치의 데이터를 추출할 수 있습니다.

* 또한 ix[ ]는 ix[0,2] 이렇게 인덱스만으로 접근을 할 수 있는데 , 칼럼 명칭기반 접근과 칼럼 위치기반 접근을 동시에 지원하기에 혼돈을 주거나 가독성이 떨어지면서 ix[ ]는 판다스에서 잘 사용하지 않습니다.

**명칭 기반 인덱싱과 위치 기반 인덱싱의 구분**

* 명칭(Label) 기반 인덱싱은 칼럼의 명칭을 기반으로 위치를 지정하는 방식입니다.

* 위치(Position) 기반 인덱싱은 0을 출발점으로 하는 가로축 , 세로축 좌표 기반의 행과 열 위치를 기반으로 데이터를 지정합니다.

* 기본적으로 DataFrame은 **명칭 기반 인덱싱**이라고 간주해야 합니다.

**DataFrame iloc[ ] 연산자**

* iloc[ ]는 위치 기반 인덱싱만 허용하기 때문에 행과 열 값으로 integer 또는 integer형의 슬라이싱 , 팬시 리스트 값을 입력해줘야 합니다.

* iloc[ ]는 슬라이싱과 팬시 인덱승은 제공하나 명확한 위치 기반 인덱싱이 사용되어야 하는 제약으로 인해 불린(boolean)인덱싱은 제공하지 않습니다.

In [30]:
train_data.head()

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S
3            4         1       1  ...  53.1000  C123         S
4            5         0       3  ...   8.0500   NaN         S

[5 rows x 12 columns]

In [36]:
train_data.iloc[3,3]

'Futrelle, Mrs. Jacques Heath (Lily May Peel)'

In [41]:
train_reset=train_data.reset_index()
train_reset.iloc[4,5]

'male'

**DataFrame loc [ ] 연산자**

* loc[ ]는 명치 기반으로 데이터를 추출합니다. 따라서 행 위치에는 DataFrame index값을 그리고 열 위치에는 칼럼 명을 입력해 줍니다.

* loc[ ]에 슬라이싱을 적용할 때는 한 가지 유의할 점이 있습니다.

* 일반적으로 슬라이싱을 '시작값' : '종료값' 과 같이 지정하면 시작값 ~ 종료값 -1 을 의미하는데

* loc[ ] 에 슬라이싱을 적용하면 종료값-1이 아닌 종료값까지 포함하는 것을 의미하는데 이는 명칭 기반 인덱싱의 특성 때문입니다

* 명칭은 숫자 형이 아닐 수 있기 때문에 -1을 할 수가 없습니다.

In [44]:
train_data.loc[0:1,'Pclass']

0    3
1    1
Name: Pclass, dtype: int64

1. 가장 중요한 것은 명칭 기반 인덱싱과 위치 기반 인덱싱의 차이를 이해하는 것입니다. DataFrame의 인덱스나 칼럼명으로 데이터에 접근 하는 것은 명칭 기반 인덱싱입니다. 0부터 시작하는 행,열의 위치 자표에만 의존하는 것이 위치 기반 인덱싱입니다.

2. ix [ ]는 명칭 기반 인덱싱과 위치 기반 인덱싱을 모두 적용할 수 있습니다. DataFrame의 인덱스가 숫자 형일 경우 행 위치에 오는 숫자는 위치 기반 인덱싱이 아니라 명칭 기반 인덱싱의 DataFrame 인덱스를 가리킵니다.

3. iloc[ ]는 위치 기반 인덱싱만 가능합니다. 따라서 행과 열 위치 값으로 정수형 값을 지정해 원하는 데이터를 반환합니다.

4. loc[ ]는 명칭 기반 인덱싱만 가능합니다. 따라서 행 위치에 DataFrame 인덱스가 오며 , 열 위치에는 칼럼명을 지정해 원하는 데이터를 반환합니다.

5. 명칭 기반 인덱싱에서 슬라이싱을 '시작점' : '종료점' 으로 지정할 때 시작점에서 종료점을 포함한 위치에 있는 데이터를 반환합니다.

**불린 인덱싱**

* 불린 인덱싱은 매우 편리한 데이터 필터링 방식입니다. 오히려 ix,iloc,loc와 같이 명확히 인덱싱을 지정하는 방식보다는 불린 인덱싱에 의존해 데이터를 가져오는 경우가 더 많습니다.

* 왜냐하면 명칭이나 위치 지정 인덱싱에서 가져올 값은 주로 로직이나 조건에 의해 계산한 뒤에 인덱싱 하는데 그럴 필요 없이 불린 인덱싱을 이용하면 원하는 값으로 필터링 하기 때문입니다.

In [47]:
train_data[train_data['Age']>60].head(5)

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
33            34         0       2  ...  10.5000   NaN         S
54            55         0       1  ...  61.9792   B30         C
96            97         0       1  ...  34.6542    A5         C
116          117         0       3  ...   7.7500   NaN         Q
170          171         0       1  ...  33.5000   B19         S

[5 rows x 12 columns]

In [48]:
train_data[train_data['Age']>60][['Name','Sex']].head(5)

Name   Sex
33            Wheadon, Mr. Edward H  male
54   Ostby, Mr. Engelhart Cornelius  male
96        Goldschmidt, Mr. George B  male
116            Connors, Mr. Patrick  male
170       Van der hoef, Mr. Wyckoff  male

* Boolean인덱싱을 할 때 여러가지의 복합 조건도 가능합니다
  * and 조건일 때는 &
  * or 조건일 때는 |
  * Not 조건일 때는 ~

In [51]:
train_data[(train_data['Age']>60) & (train_data['Pclass']==1) & (train_data['Sex']=='female')]

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
275          276         1       1  ...  77.9583    D7         S
829          830         1       1  ...  80.0000   B28       NaN

[2 rows x 12 columns]

## 7. 정렬 , Aggregation 함수 , GroupBy 적용

**DataFrame , Series의 정렬 - sort_values()**

* DataFrame과 Series의 정렬을 위해서는 sort_values() 메소드를 이용합니다.

* sort_values()의 주요 입력 파라미터는 **by , ascending , inplace**입니다.

* by로 특정 칼럼을 입력하면 해당 칼럼으로 정렬을 수행합니다.

* ascending=True로 설정하면 오름차순으로 정렬하며 , False로 설정하면 내림차순으로 정렬 합니다.

* inplace=False로 설정하면 sort_value를 호출한 기존 DataFrame은 그대로 유지되며 정렬된 DataFrame을 반환합니다.

In [52]:
train_sorted=train_data.sort_values(by=['Name'])
train_sorted.head(5)

PassengerId  Survived  Pclass  ...   Fare Cabin  Embarked
845          846         0       3  ...   7.55   NaN         S
746          747         0       3  ...  20.25   NaN         S
279          280         1       3  ...  20.25   NaN         S
308          309         0       2  ...  24.00   NaN         C
874          875         1       2  ...  24.00   NaN         C

[5 rows x 12 columns]

* 여러 개의 칼럼으로 정렬하려면 by에 리스트 형식으로 정렬하려는 칼럼을 입력하면 됩니다.

In [ ]:
train_sorted=train_data.sort_values(by=['Pclass','Name'])
train_sorted.head(5)

**Aggregation함수 적용**

* DataFrame에서 min(),max(),sum(),count()와 같은 aggregation 함수의 적용은 모든 칼럼에 해당 aggregation을 적용합니다.

* train_data 에 count()를 적용하면 모든 칼럼에 count()결과를 반환합니다.

In [54]:
train_data.count()

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64

* 특정 칼럼에 aggregation 함수를 적용하기 위해서는 DataFrame에 대상 칼럼들만 추출해 aggregation을 적용하면 됩니다.

In [56]:
train_data[['Age','Fare']].mean()

Age     29.699118
Fare    32.204208
dtype: float64

**groupby() 적용**

* 판다스는 groupby를 분석 작업에 매우 많이 활용합ㅂ니다.

* DataFrame의 groupby() 사용 시 입력 파라미터 by에 칼럼을 입력하면 대상 칼럼으로 groupby됩니다.

* DataFrame에 groupby()를 호출하면 DataFrameGroupBy라는 또 다른 형태의 DataFrame을 반환합니다.

* groupby(by='Pclass')를 호출하면 Pclass 칼럼 기준으로 GroupBy된 DataFrame Groupby 객체를 반환합니다.

In [58]:
train_groupby=train_data.groupby(by='Pclass')
print(type(train_groupby))

<class 'pandas.core.groupby.generic.DataFrameGroupBy'>


* DataFrame에 groupby()를 호출해 반환된 결과에 aggregation 함수를 호출하면 groupby() 대상 칼럼을 제외한 모든 칼럼에 해당 aggregation 함수를 적용합니다.


In [59]:
train_groupby.count() # Pclass 별로 count 결과를 나타내줌

PassengerId  Survived  Name  Sex  ...  Ticket  Fare  Cabin  Embarked
Pclass                                    ...                               
1               216       216   216  216  ...     216   216    176       214
2               184       184   184  184  ...     184   184     16       184
3               491       491   491  491  ...     491   491     12       491

[3 rows x 11 columns]

## 8. 결손 데이터 처리하기

* 판다스는 결손 데이터(Missing Data)를 처리하는 편리한 API를 제공합니다.

* 결손 데이터는 칼럼에 값이 없는 , 즉 NULL인 경우를 의미하며 , 이를 넘파이의 NaN으로 표시합니다.

* 기본적으로 머신러닝 알고리즘은 이 NaN 값을 처리하지 않으므로 이 값을 다른 값으로 대체해야 합니다.

* 또한 NaN 값은 평균 , 총합 등의 함수 연산 시 제외가 됩니다.

* NaN 여부를 확인하는 API는 isna() 이며 , NaN 값을 다른 값으로 대체하는 API는 fillna()입니다.

**isna()로 결손 데이터 여부 확인**

* isna()는 데이터가 NaN인지 아닌지를 알려줍니다. DataFrame에 isna()를 수행하면 모든 칼럼의 값이 NaN인지 아닌지를 True나 False로 알려줍니다.

* 결손 데이터의 갯수는 isna() 결과에 sum()함수를 추가해 구할 수 있습니다.

In [60]:
train_data.isna().head(5)

PassengerId  Survived  Pclass   Name  ...  Ticket   Fare  Cabin  Embarked
0        False     False   False  False  ...   False  False   True     False
1        False     False   False  False  ...   False  False  False     False
2        False     False   False  False  ...   False  False   True     False
3        False     False   False  False  ...   False  False  False     False
4        False     False   False  False  ...   False  False   True     False

[5 rows x 12 columns]

In [61]:
train_data.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

**fillna()로 결손 데이터 대체하기**

* fillna()를 이용하면 결손 데이터를 편리하게 다른 값으로 대체할 수 있습니다.

* 주의해야 할 점은 fillna()를 이용해 반환 값을 다시 받거나 inplace=True 파라미터를 fillna()에 추가해야 실제 데이터 세트 값이 변경된다는 점 입니다.

* 타이타닉 데이터 세트의 'Cabin' 칼럼의 NaN값을 'C000'으로 대체해보겠습니다.

In [63]:
train_data['Cabin']=train_data['Cabin'].fillna('C000')

train_data.head(5)

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500  C000         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250  C000         S
3            4         1       1  ...  53.1000  C123         S
4            5         0       3  ...   8.0500  C000         S

[5 rows x 12 columns]

* 'Age' 칼럼의 NaN 값을 평균 나이로 , 'Embarked' 칼럼의 NaN 값을 'S'로 대체해 모든 결손 데이터를 처리하겠습니다.

In [65]:
train_data['Age']=train_data['Age'].fillna(train_data['Age'].mean())
train_data['Embarked']=train_data['Embarked'].fillna('S')
## 모든 결손 데이터를 처리해줬음
train_data.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

## 9. apply lambda 식으로 데이터 가공

* 판다스는 apply 함수에 lambda 식을 결합해 DataFrame이나 Series의 레코드별로 데이터를 가공하는 기능을 제공합니다.

* 판다스의 경우 칼럼에 일괄적으로 데이터 가공을 하는 것이 속도 면에서 더 빠르나 복잡한 데이터 가공이 필요할 경우 어쩔 수 없이 apply lambda를 이용합니다.

* lambda 식은 파이썬에서 함수형 프로그래밍을 지원하기 위해 만들었습니다.

* 가령 다음과 같이 입력값의 제곱 값을 구해서 반환하는 get_square(a)라는 함수가 있다고 가정해보면

* lambda 함수는 이러한 함수의 선언과 함수 내의 처리를 한 줄의 식으로 쉽게 변환하는 식입니다.

In [68]:
def get_square(a):
  return a**2

print("[def] 3의 제곱은 : {}".format(get_square(3)))

lambda_square=lambda x : x**2
print("[lambda] 3의 제곱은 : {}".format(lambda_square(3)))


[def] 3의 제곱은 : 9
[lambda] 3의 제곱은 : 9


* lambda 식을 이용할 때 여러 개의 값을 입력 인자로 사용해야 할 경우에는 보통 map() 함수를 결합해서 사용합니다.

In [69]:
a=[1,2,3]
squares=map(lambda x : x**2,a)

list(squares)

[1, 4, 9]

* 판다스 DataFrame의 lambda 식은 파이썬의 이러한 lambda 식을 그대로 적용한 것입니다.

* 먼저 'Name' 칼럼의 문자열 개수를 별도의 칼럼인 'Name_len'에 생성해 보겠습니다.

In [70]:
train_data['Name_len']=train_data['Name'].apply(lambda x : len(x))
train_data[['Name','Name_len']].head(3)

Name  Name_len
0                            Braund, Mr. Owen Harris        23
1  Cumings, Mrs. John Bradley (Florence Briggs Th...        51
2                             Heikkinen, Miss. Laina        22

* Lambda 식에서 if else 절을 사용해 조금 더 복잡한 가공을 해보겠습니다.

* 나이가 15세 미만이면 'Child' , 그렇지 않으면 'Adult'로 구분하는 새로운 칼럼 'Child_Adult'를 apply lambda를 이용해 만들어 보겠습니다.

In [72]:
train_data['Child_Adult']=train_data['Age'].apply(lambda x : 'Child' if x<=15 else 'Adult')
train_data[['Age','Child_Adult']].head(3)

Age Child_Adult
0  22.0       Adult
1  38.0       Adult
2  26.0       Adult

* lambda 식은 if else를 지원하는데 , 주의할 점이 있습니다.

* if 절의 경우 if 식보다 반환 값을 먼저 기술 해주고 if 구문을 사용해야 합니다. lambda 식 : 기호의 오른편에는 반환 값이 있어야 하기 때문입니다.

* if , else 는 지원하지만 else if는 지원하지 않습니다.

# 정리

* 넘파이와 판다스를 완벽하게 정복하는 것은 너무 광범위 하기 때문에 힘들다

* 따라서 넘파이와 판다스는 머신러닝 프로그램을 직접 만들면서 이해하고 , 모르는 부분들을 인터넷 검색 등으로 찾아 보는 게 학습을 위한 제일 빠른 방법이다.

* 앞으로 공부하면서 느끼겠지만 , 실제로 Machine Learning 모델을 생성하고 예측을 수행하는 데 있어서 ML 알고리즘이 차지하는 비중보다 데이터를 전처리하고 적절한 피처를 가공/추출하는 부분이 훨씬 많은 비중을 차지하게 됩니다.

* 넘파이 , 판다스 , 맷플롭립/시본 과 같이 파이썬 머신러닝 생태계를 이루고 있는 다양한 패키지의 이해는 그렇기 때문에 매우 중요합니다.